### PR TITLE
Codata streams by coinductive records

### PR DESCRIPTION
**Pull request by Taneb**
I'm using this for a bunch of personal projects so I'd like to get it in stdlib. However, it's nowhere near ready, really.

I'm only defining Streams for now because they have a regular structure, I don't need to worry about Maybes like I would with Colists or Conats defined this way.

I don't know how far I want to take various relations and properties. "All the way" might be the right answer for this.

I tried to be at least as complete as the other two definitions of streams, but I couldn't figure out how to implement some things, such as `interleave⁺ : List⁺ (Stream A) → Stream A`

### CODE REVIEW

File: src/Codata/Guarded/Stream/Relation/Binary/Pointwise.agda [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+open import Codata.Guarded.Stream as Stream using (Stream; head; tail)
+open import Data.Nat.Base using (ℕ)
+open import Function.Base using (_∘_)
+open import Level
+open import Relation.Binary
+
+private
+  variable
+    a b c d ℓ : Level
+    A : Set a
```
Comment (MatthewDaggitt) [inline src/Codata/Guarded/Stream/Relation/Binary/Pointwise.agda:14]: Could define this as `A B C D : Set a`

File: src/Codata/Guarded/Stream.agda [hunk 2]
```diff
@@ -0,0 +1,138 @@
+{-# OPTIONS --safe --without-K --guardedness #-}
+------------------------------------------------------------------------
```
Comment (MatthewDaggitt) [inline src/Codata/Guarded/Stream.agda:2]: Missing proper standard library module headers. Also the pragmas should go underneath the module headers.

**Review by MatthewDaggitt — COMMENTED** [review]
See https://github.com/agda/agda-stdlib/issues/1588 for related discussion.

This looks fine to me! Happy to merge things in piece wise if you'd like?

**Comment by gallais** [comment]
This should work:

```agda
interleave⁺ : List⁺ (Stream A) → Stream A
interleave⁺ {A = A} (x ∷ xs) = go [] x xs where

  go : List (Stream A) → Stream A → List (Stream A) → Stream A
  head (go rev xs yss) = head xs
  tail (go rev xs [])         = interleave⁺ (List⁺.reverse (tail xs ∷ rev))
  tail (go rev xs (ys ∷ yss)) = go (tail xs ∷ rev) ys yss
```

File: src/Codata/Guarded/Stream.agda [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+--
+-- Infinite streams defined as coinductive records
+------------------------------------------------------------------------
+
+{-# OPTIONS --safe --without-K --guardedness #-}
+
+module Codata.Guarded.Stream where
+
+open import Data.List.Base using (List)
+open import Data.List.NonEmpty.Base as List⁺ using (List⁺)
```
Comment (gallais) [inline src/Codata/Guarded/Stream.agda:12]:
```suggestion
open import Data.List.Base using (List; []; _∷_)
open import Data.List.NonEmpty.Base as List⁺ using (List⁺; _∷_)
```
Agda does type-directed disambiguation of constructors so you can import all of these and use them unqualified
in the module.

**Comment by Taneb** [comment]
I started implementing non-empty colists like

```agda
record Colist+ (A : Set a) : Set a where
  coinductive
  field
    head : a
    tail : Maybe (Colist+ A)
```

and while it's _possible_ to do things that way, it's frustrating that the termination checker forces me to pattern-match on the Maybe in the tail a lot, I can't just use (e.g.) `Maybe.map (map f) (tail xs)`

**Comment by MatthewDaggitt** [comment]
What's the plan for this? Happy to merge it in if a changelog entry gets added?

**Comment by Taneb** [comment]
Yeah. I'll try to get a changelog entry sorted today

**Review by gallais — APPROVED** [review]

**Pull request merged**